### PR TITLE
Restructure library with  zne/ subpackage to close issue #293

### PR DESCRIPTION
**Pull request by nathanshammah**
Description
-----------
Restructure library to close issue #293:  
- [x] Add `zne/` subpackage 
- [x] Rename `factories.py` into `inference.py` 
- [x] Rename `folding.py` into `scaling.py`
- [ ] Add shortcuts in init file(s), e.g., allowing
```python
from mitiq.inference import RichardsonFactory
```
for
```python
from mitiq.zne.inference import RichardsonFactory
```

For consitency:
- [x] Fix imports in library
- [x] Fix references to structure in README 
- [x] Fix references to structure in documentation

Checklist
-----------
Please make sure you have finished the following tasks before requesting a review of the pull request (PR). For more information, please refer to the [Contributor's Guide](CONTRIBUTING.md):

- [x] Contributions to mitiq should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/). You can enforce it easily with [`black`][black] style and with [`flake8`][flake8] conventions.
- [x] Please add tests to cover your changes, if applicable, and make sure that all new and existing tests pass locally.
- [x] If the behavior of the code has changed or new feature has been added, please also [update the documentation](docs/README-docs.md).
- [x] Functions and classes have useful [Google-style][google] docstrings and [type hints][hints] in the signature of the objects.
- [x] (Bug fix) The associated issue is referenced above using [auto-close keywords][auto-close]. If the PR fixes an issue, use the keyword fix/fixes/fixed followed by the issue id. For example, if the PR fixes issue 1184, type "Fixes #1184" (without quotes).
- [x] The [changelog][changelog] is updated, including author and PR number (@username, gh-xxx).

[auto-close]: https://help.github.com/en/articles/closing-issues-using-keywords
[black]: https://black.readthedocs.io/en/stable/index.html
[changelog]: https://github.com/unitaryfund/mitiq/blob/master/CHANGELOG.md
[contributing]: https://github.com/unitaryfund/mitiq/blob/master/CONTRIBUTING.md
[docs]: https://github.com/unitaryfund/mitiq/blob/master/docs/source/
[flake8]: http://flake8.pycqa.org
[google]: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
[hints]: https://www.python.org/dev/peps/pep-0484/
[pep-484]: https://www.python.org/dev/peps/pep-0484/

### CODE REVIEW

**Comment by nathanshammah** [comment]
@rmlarose this is ready for review. I have not implemented one of the bullet points above, the shortcuts (there is some ugly `mitiq.zne.zne` module) just because I am not sure of the way we want it:

If we import aliases in one of the init (top level or in `mitiq.zne`), every time one is importing the top module or zne, one is importing all...?


```
# in the top __init__.py file:
import mitiq.zne.inference as mitiq.inference
```
we don't want this, do we? as it would slow things down on start-up. We just want that 
```
from mitiq.inference import *
```
**by the user** maps to `import mitiq.zne.inference`, is it correct?

**Comment by nathanshammah** [comment]
@andreamari, I updated this as discussed in the meeting, ready for review.

File: docs/source/guide/guide_05-factories.rst [hunk 1]
```diff
@@ -19,39 +19,39 @@ The typical tasks of a factory are:
...
-Once instantiated, a factory can be passed as an argument to the high-level functions contained in the module :py:mod:`mitiq.zne`.
-Alternatively, a factory can be directly used to implement a zero-noise extrapolation procedure in a fully self-contained way. 
+Once instantiated, a factory can be passed as an argument to the high-level functions contained in the module :py:mod:`mitiq.zne.zne`.
+Alternatively, a factory can be directly used to implement a zero-noise extrapolation procedure in a fully self-contained way.
 
 To clarify this aspect, we now perform the same zero-noise extrapolation with both methods.
 
 ===============================================================
-Using a factory object with the :py:mod:`mitiq.zne` module
+Using a factory object with the :py:mod:`mitiq.zne.zne` module
```
Comment (rmlarose) [inline docs/source/guide/guide_05-factories.rst:54]:
```suggestion
Using a factory object with the :py:mod:`mitiq.zne` module
```

Comment (rmlarose) [reply docs/source/guide/guide_05-factories.rst:54]: @nathanshammah `mitiq.zne` is the module no?

File: docs/source/guide/guide_06-error-mitigation.rst [hunk 2]
```diff
@@ -46,7 +46,7 @@ then re-running the calculation (which is indeed a time evolution) for
 :math:`\langle X\rangle_{0}`.
 The extraction for :math:`\langle X\rangle_{0}` can occur with several
 statistical fitting models, which can be linear or non-linear. These methods
-are contained in the :mod:`mitiq.factories` and :mod:`mitiq.zne` modules.
+are contained in the :mod:`mitiq.zne.inference` and :mod:`mitiq.zne.zne` modules.
```
Comment (rmlarose) [inline docs/source/guide/guide_06-error-mitigation.rst:49]:
```suggestion
are contained in the :mod:`mitiq.zne.inference` and :mod:`mitiq.zne` modules.
```

File: mitiq/zne/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Licensed under GNI GPL v.3.0
+"""
+This is the mitiq.zne subpackage from which functions and classes of
+Mitiq can be directly imported.
+"""
```
Comment (rmlarose) [inline mitiq/zne/__init__.py:5]:
```suggestion
"""
from mitiq.zne.zne import execute_with_zne, mitigate_executor, zne_decorator
```

File: mitiq/mitiq_qiskit/tests/test_mitiq_qiskit_zne.py [hunk 4]
```diff
@@ -5,8 +5,8 @@
 
 from qiskit import ClassicalRegister, QuantumCircuit
 
-from mitiq.factories import RichardsonFactory, ExpFactory
-from mitiq.zne import (
+from mitiq.zne.inference import RichardsonFactory, ExpFactory
+from mitiq.zne.zne import (
```
Comment (rmlarose) [inline mitiq/mitiq_qiskit/tests/test_mitiq_qiskit_zne.py:9]:
```suggestion
from mitiq.zne import (
```

File: mitiq/mitiq_pyquil/tests/test_zne.py [hunk 5]
```diff
@@ -2,8 +2,8 @@
 
 import numpy as np
 from mitiq._typing import QPROGRAM
-from mitiq.factories import RichardsonFactory
-from mitiq.zne import (
+from mitiq.zne.inference import RichardsonFactory
+from mitiq.zne.zne import (
```
Comment (rmlarose) [inline mitiq/mitiq_pyquil/tests/test_zne.py:6]:
```suggestion
from mitiq.zne import (
```

File: mitiq/benchmarks/tests/test_randomized_benchmaking.py [hunk 6]
```diff
@@ -3,21 +3,21 @@
...
-from mitiq.folding import (
+from mitiq.zne.scaling import (
     fold_gates_at_random,
     fold_gates_from_left,
     fold_gates_from_right,
     fold_global,
 )
 from mitiq.benchmarks.utils import noisy_simulation
-from mitiq.zne import mitigate_executor
+from mitiq.zne.zne import mitigate_executor
```
Comment (rmlarose) [inline mitiq/benchmarks/tests/test_randomized_benchmaking.py:20]:
```suggestion
from mitiq.zne import mitigate_executor
```

File: mitiq/__init__.py [hunk 7]
```diff
@@ -7,4 +7,4 @@
 from mitiq._about import about
 from mitiq._typing import QPROGRAM
 from mitiq._version import __version__
-from mitiq.zne import execute_with_zne, mitigate_executor
+from mitiq.zne.zne import execute_with_zne, mitigate_executor, zne_decorator
```
Comment (rmlarose) [inline mitiq/__init__.py:10]:
```suggestion
from mitiq.zne import execute_with_zne, mitigate_executor, zne_decorator
```

File: docs/source/guide/guide_08-qaoa-landscape.rst [hunk 8]
```diff
@@ -15,7 +15,7 @@ alternating operator ansatz (QAOA). We first import the libraries we will use.
     import numpy as np
 
     from cirq import Circuit, CNOT, DensityMatrixSimulator, H, LineQubit, depolarize, rz
-    from mitiq.zne import mitigate_executor
+    from mitiq.zne.zne import mitigate_executor
```
Comment (rmlarose) [inline docs/source/guide/guide_08-qaoa-landscape.rst:18]:
```suggestion
    from mitiq.zne import mitigate_executor
```

File: mitiq/zne/tests/test_zne.py [hunk 9]
```diff
@@ -5,13 +5,13 @@
...
-from mitiq.factories import LinearFactory, RichardsonFactory
-from mitiq.folding import (
+from mitiq.zne.inference import LinearFactory, RichardsonFactory
+from mitiq.zne.scaling import (
     fold_gates_from_left,
     fold_gates_from_right,
     fold_gates_at_random,
 )
-from mitiq.zne import execute_with_zne, mitigate_executor, zne_decorator
+from mitiq.zne.zne import execute_with_zne, mitigate_executor, zne_decorator
```
Comment (rmlarose) [inline mitiq/zne/tests/test_zne.py:14]:
```suggestion
from mitiq.zne import execute_with_zne, mitigate_executor, zne_decorator
```

File: docs/source/guide/guide_05-factories.rst (hunk 1 above)
Comment (nathanshammah) [reply docs/source/guide/guide_05-factories.rst:54]: mitiq.zne is the package.

**Comment by andreamari** [comment]
> @nathanshammah I think we definitely want to have
> 
> ```python
> from mitiq.zne.zne import execute_with_zne, mitigate_executor, zne_decorator
> ```
> 
> in `mitiq/zne/__init__.py`. This gets rid of the awkward `from mitiq.zne.zne import` in favor of `from mitiq.zne import`.

+1

**Pull request merged**